File: TMVA/create_skimmed_ntuple.c
#include <iostream>
#include <TTree.h>
#include <TFile.h>
#include <TCanvas.h>
#include <TGraph.h>


///////////////////
//     MAIN      //
///////////////////
void create_skimmed_ntuple()
{
  // Get name of the TMVA method
  TString method_name = "";
  ifstream config_file("tmva_config.txt", ifstream::binary);
  if (config_file.is_open()) {
    string str1;
    string delim = ": ";
    while(getline(config_file, str1)) {
      string par_name = str1.substr(0, str1.find(delim));
      string par_val  = str1.substr(str1.find(delim)+2, str1.size()); // +2 due to delim length
      if (par_name == "TMVA Method")  method_name += par_val;
    }
  }
  config_file.close();



  // Set some variables for the analysis
  double classifier_cut = 0.1;
  int n_jets_from_top = 0;
  map<int, map<int, vector<double>> > event_jet_jetinfo;
  // { {event_0: {jet_0 - {DL1r, KBDT, etc}}, {jet_1 - {DL1r, KBDT, etc}} ... },
  //   {event_1: {jet_0 - {DL1r, KBDT, etc}}, {jet_1 - {DL1r, KBDT, etc}} ... }, etc }
  map<int, vector<double>> event_lepinfo; // + weight and met!
  // { {event_0: {weight, met, el_pt, el_eta, el_phi, el_e, el_charge, mu_pt, ..},
  //   {event_1: {weight, met, el_pt, el_eta, el_phi, el_e, el_charge, mu_pt, ..}, etc }


  // Open the tfile - mtva output
  TFile *tmva_file = new TFile("TMVA.root");
  
  
  // Open trees
  vector<TString> tree_names = {"TestTree", "TrainTree"};
  for (int tree_count=0; tree_count<tree_names.size(); tree_count++) {
    TTree *tmva_tree = (TTree*)tmva_file->Get("dataset/" + tree_names[tree_count]);
    cout << "working with " << tree_names[tree_count] << " tree!" << endl;
    
    
    // Set branches
    float topHOF, jet_isbtagged_DL1r_77, jet_DL1r, classifier, weight, jet_pt, jet_eta, jet_phi, jet_e;
    float mu_pt, mu_eta, mu_phi, mu_e, mu_charge;
    float el_pt, el_eta, el_phi, el_e, el_charge; 
    float event_number, met;
    topHOF = jet_isbtagged_DL1r_77 = jet_DL1r = weight = jet_pt = jet_eta = jet_phi, jet_e = 0;
    mu_pt = mu_eta = mu_phi = mu_e = mu_charge = 0;
    el_pt = el_eta = el_phi = el_e = el_charge = 0;
    event_number = met = 0;

    tmva_tree->SetBranchAddress("topHadronOriginFlag", &topHOF);
    tmva_tree->SetBranchAddress("jet_isbtagged_DL1r_77", &jet_isbtagged_DL1r_77);
    tmva_tree->SetBranchAddress("NN_jet_DL1r", &jet_DL1r);
    tmva_tree->SetBranchAddress(method_name, &classifier);
    tmva_tree->SetBranchAddress("NN_tot_event_weight", &weight);
    tmva_tree->SetBranchAddress("NN_jet_pt", &jet_pt);
    tmva_tree->SetBranchAddress("NN_jet_eta", &jet_eta);
    tmva_tree->SetBranchAddress("NN_jet_phi", &jet_phi);
    tmva_tree->SetBranchAddress("NN_jet_e", &jet_e);
    tmva_tree->SetBranchAddress("NN_mu_pt", &mu_pt);
    tmva_tree->SetBranchAddress("NN_mu_eta", &mu_eta);
    tmva_tree->SetBranchAddress("NN_mu_phi", &mu_phi);
    tmva_tree->SetBranchAddress("NN_mu_e", &mu_e);
    tmva_tree->SetBranchAddress("NN_mu_charge", &mu_charge);
    tmva_tree->SetBranchAddress("NN_el_pt", &el_pt);
    tmva_tree->SetBranchAddress("NN_el_eta", &el_eta);
    tmva_tree->SetBranchAddress("NN_el_phi", &el_phi);
    tmva_tree->SetBranchAddress("NN_el_e", &el_e);
    tmva_tree->SetBranchAddress("NN_el_charge", &el_charge);
    tmva_tree->SetBranchAddress("NN_event_number", &event_number);
    tmva_tree->SetBranchAddress("NN_met", &met);

    
    // Get number of entries
    Int_t nEntries = tmva_tree->GetEntries();
    cout << "nEntries: " << nEntries << endl << endl;
    
  
    for (int entry=0; entry<nEntries; entry++) {
      
      if (entry%1000==0) { cout << "\t" << entry << "\r"; cout.flush(); }
      tmva_tree->GetEntry(entry);
   
    
      // Black magic
      // jets info
      map<int, vector<double> > jet_jetinfo;
      vector<double> jetinfo = {topHOF, jet_isbtagged_DL1r_77, jet_DL1r, classifier, jet_pt, jet_eta, jet_phi, jet_e};
      int jetn;
    
      map<int, map<int, vector<double>> >::iterator it1 = event_jet_jetinfo.find(event_number);
      if ( it1 != event_jet_jetinfo.end() ) {
	jet_jetinfo = it1->second;
	jetn = jet_jetinfo.size();
	jet_jetinfo.insert(make_pair(jetn, jetinfo));
	it1->second = jet_jetinfo; }
      else {
	jet_jetinfo.insert(make_pair(0, jetinfo));
	event_jet_jetinfo.insert(make_pair(event_number, jet_jetinfo)); }
      
      // leptons info + met + weight
      vector<double> lepinfo = {weight, met, el_pt, el_eta, el_phi, el_e, el_charge, mu_pt, mu_eta, mu_phi, mu_e, mu_charge};
      map<int, vector<double>>::iterator it2 = event_lepinfo.find(event_number);
      if ( event_lepinfo.count(event_number) == 0 ) event_lepinfo.insert(make_pair(event_number, lepinfo));
      
    } // [entry] - loop over entries
    
  } // [tree_count] - loop over trees
    
  tmva_file->Close();


  // ///
  // Work with the map of events
  // ///

  int n_jets_in_map_map = 0;
  
  // Create a new TFile and TTree
  TFile *ntuple = new TFile("skimmed_test_ntuple.root", "RECREATE");
  TTree *tree = new TTree("test", "Tree with entries from test tmva tree");

  // Declare variables for a tree
  int event_number_save;
  vector<float> topHOF_save, jet_isbtagged_DL1r_77_save, jet_DL1r_save, classifier_save, jet_pt_save, jet_eta_save, jet_phi_save, jet_e_save;
  float weight_save;
  float mu_pt_save, mu_eta_save, mu_phi_save, mu_e_save, mu_charge_save;
  float el_pt_save, el_eta_save, el_phi_save, el_e_save, el_charge_save;
  float met_save;
  event_number_save = 0;

  // Declare new branches
  tree->Branch("topHadronOriginFlag", &topHOF_save);
  tree->Branch("jet_isbtagged_DL1r_77", &jet_isbtagged_DL1r_77_save);
  tree->Branch("jet_DL1r", &jet_DL1r_save);
  tree->Branch(method_name, &classifier_save);
  tree->Branch("tot_event_weight", &weight_save, "tot_event_weight/F");
  tree->Branch("jet_pt", &jet_pt_save);
  tree->Branch("jet_eta", &jet_eta_save);
  tree->Branch("jet_phi", &jet_phi_save);
  tree->Branch("jet_e", &jet_e_save);
  tree->Branch("mu_pt", &mu_pt_save);
  tree->Branch("mu_eta", &mu_eta_save);
  tree->Branch("mu_phi", &mu_phi_save);
  tree->Branch("mu_e", &mu_e_save);
  tree->Branch("mu_charge", &mu_charge_save);
  tree->Branch("el_pt", &el_pt_save);
  tree->Branch("el_eta", &el_eta_save);
  tree->Branch("el_phi", &el_phi_save);
  tree->Branch("el_e", &el_e_save);
  tree->Branch("el_charge", &el_charge_save);
  tree->Branch("event_number", &event_number_save, "event_number/I");
  tree->Branch("met", &met_save, "met/F");
  

  // Run a loop over keys - event numbers
  map<int, map<int, vector<double>> >::iterator it1;
  for (it1 = event_jet_jetinfo.begin(); it1 != event_jet_jetinfo.end(); it1++) {

    // Work out jets info
    topHOF_save.clear();
    jet_isbtagged_DL1r_77_save.clear();
    jet_DL1r_save.clear();
    classifier_save.clear();
    jet_pt_save.clear();
    jet_eta_save.clear();
    jet_phi_save.clear();
    jet_e_save.clear();
    
    //cout << "\n" << it1->first << endl;

    // Set a map for jets=keys and properties=values for the event
    map<int, vector<double> > jet_jetinfo = it1->second;
    
    // Run a loop over keys of one event - jets (that are not in any reasonable order)
    map<int, vector<double> >::iterator it2;
    for (it2 = jet_jetinfo.begin(); it2 != jet_jetinfo.end(); it2++) {
      int jetn = it2->first;
      vector<double> jet_pars = it2->second;
      n_jets_in_map_map++;
      
      //cout << "\tjet_# = " << jetn << ";\t dl1r = " << jet_pars[0] << ";\t classifier = " << jet_pars[1] << ";\tweight = " << jet_pars[2] << ";\ttopHOF = " << jet_pars[3] << endl; 

      topHOF_save.push_back(jet_pars[0]);
      jet_isbtagged_DL1r_77_save.push_back(jet_pars[1]);
      jet_DL1r_save.push_back(jet_pars[2]);
      classifier_save.push_back(jet_pars[3]);
      jet_pt_save.push_back(jet_pars[4]);
      jet_eta_save.push_back(jet_pars[5]);
      jet_phi_save.push_back(jet_pars[6]);
      
    } // [it2] - loop over jets of an event
   

    // Work out leptons info, met and event number
    event_number_save = it1->first;
    weight_save = (event_lepinfo[event_number_save])[0];
    met_save = (event_lepinfo[event_number_save])[1];
    el_pt_save = (event_lepinfo[event_number_save])[2];
    el_eta_save = (event_lepinfo[event_number_save])[3];
    el_phi_save = (event_lepinfo[event_number_save])[4];
    el_e_save = (event_lepinfo[event_number_save])[5];
    el_charge_save = (event_lepinfo[event_number_save])[6];
    mu_pt_save = (event_lepinfo[event_number_save])[7];
    mu_eta_save = (event_lepinfo[event_number_save])[8];
    mu_phi_save = (event_lepinfo[event_number_save])[9];
    mu_e_save = (event_lepinfo[event_number_save])[10];
    mu_charge_save = (event_lepinfo[event_number_save])[11];

    
    tree->Fill();
    
    // Release some space
    event_jet_jetinfo.erase(it1);
    
  } // [it1] - loop over events
  

  tree->Write("test", TTree::kOverwrite);
 

}

File: TMVA/create_skimmed_ntuple.py
import ROOT
from array import array


##################
##     MAIN     ##
##################

# Get name of the TMVA method
method_name = ""
config_file = open("tmva_config.txt", "r")
for line in config_file:
  par_name_and_var = line.split(": ")    
  if ("TMVA Method" in line):
    method_name += par_name_and_var[1].strip() # remove the \n
config_file.close()
print(method_name)


# Set some variables for the analysis
classifier_cut = 0.1
n_jets_from_top = 0
event_jet_jetinfo = {}
event_lepinfo = {}


# Open the tfile - tmva ouptut
tmva_file = ROOT.TFile.Open("TMVA.root", "READ")


# Open trees
tree_names = ["TestTree", "TrainTree"]
for tree_name in tree_names:
  tree = tmva_file.Get("dataset/" + tree_name)
  print("Working with ", tree_name, " tree!")

  # Gen number of entries
  nEntries = tree.GetEntries()
  print("nEntries: ", nEntries)

  # Run a loop over entries=jets
  for entry in range(0, nEntries):
    entry = int(entry)
  
    if (entry%1000==0): 
      print("\t", entry, end="\r")

    tree.GetEntry(entry)
      
    # Declare single values
    event_number          = getattr(tree, "event_number")
    
    weight                = getattr(tree, "tot_event_weight")
    met                   = getattr(tree, "met")
    el_pt                 = getattr(tree, "el_pt")
    el_eta                = getattr(tree, "el_eta")
    el_phi                = getattr(tree, "el_phi")
    el_e                  = getattr(tree, "el_e")
    el_charge             = getattr(tree, "el_charge")
    mu_pt                 = getattr(tree, "mu_pt")
    mu_eta                = getattr(tree, "mu_eta")
    mu_phi                = getattr(tree, "mu_phi")
    mu_e                  = getattr(tree, "mu_e")
    mu_charge             = getattr(tree, "mu_charge")

    # Declate "vectors-to-be" values
    topHOF                = getattr(tree, "topHadronOriginFlag")
    jet_isbtagged_DL1r_77 = getattr(tree, "jet_isbtagged_DL1r_77")
    jet_truthflav         = getattr(tree, "jet_truthflav")
    jet_DL1r              = getattr(tree, "jet_DL1r")
    classifier            = getattr(tree, method_name)
    jet_pt                = getattr(tree, "jet_pt")
    jet_eta               = getattr(tree, "jet_eta")
    jet_phi               = getattr(tree, "jet_phi")
    jet_e                 = getattr(tree, "jet_e")
    
    jetinfo = [topHOF, jet_isbtagged_DL1r_77, jet_truthflav, jet_DL1r, classifier, jet_pt, jet_eta, jet_phi, jet_e]

    if event_number in event_jet_jetinfo:
      jetn = len((event_jet_jetinfo[event_number]))
      jet_jetinfo = event_jet_jetinfo[event_number]
      jet_jetinfo[jetn] = jetinfo
      event_jet_jetinfo[event_number] = jet_jetinfo
    else:
      jet_jetinfo= {}
      jet_jetinfo[0] = jetinfo
      event_jet_jetinfo[event_number] = jet_jetinfo

    lepinfo = [weight, met, el_pt, el_eta, el_phi, el_e, el_charge, mu_pt, mu_eta, mu_phi, mu_e, mu_charge]

    if event_number not in event_lepinfo:
      event_lepinfo[event_number] = lepinfo

# Close the tmva file
tmva_file.Close()


# ###
# Create the new ntuple
# ###

# Create the file and a tree
new_file = ROOT.TFile.Open("skimmed_test_ntuple.root", "RECREATE")
new_tree = ROOT.TTree("test", "Tree with entries from test tmva tree")

# Create branches for "single" values per event
weight_save    = array("f", [0.0])
met_save       = array("f", [0.0])
el_pt_save     = array("f", [0.0])
el_eta_save    = array("f", [0.0])
el_phi_save    = array("f", [0.0])
el_e_save      = array("f", [0.0])
el_charge_save = array("f", [0.0])
mu_pt_save     = array("f", [0.0])
mu_eta_save    = array("f", [0.0])
mu_phi_save    = array("f", [0.0])
mu_e_save      = array("f", [0.0])
mu_charge_save = array("f", [0.0])

new_tree.Branch("tot_event_weight", weight_save,    "tot_event_weight/F")
new_tree.Branch("met",              met_save,       "met/F")
new_tree.Branch("el_pt",            el_pt_save,     "el_pt/F")
new_tree.Branch("el_eta",           el_eta_save,    "el_eta/F")
new_tree.Branch("el_phi",           el_phi_save,    "el_phi/F")
new_tree.Branch("el_e",             el_e_save,      "el_e/F")
new_tree.Branch("el_charge",        el_charge_save, "el_charge/F")
new_tree.Branch("mu_pt",            mu_pt_save,     "mu_pt/F")
new_tree.Branch("mu_eta",           mu_eta_save,    "mu_eta/F")
new_tree.Branch("mu_phi",           mu_phi_save,    "mu_phi/F")
new_tree.Branch("mu_e",             mu_e_save,      "mu_e/F")
new_tree.Branch("mu_charge",        mu_charge_save, "mu_charge/F")

# Create branches for "vector" values per event
topHOF_save                = ROOT.vector("int")()
jet_isbtagged_DL1r_77_save = ROOT.vector("int")()
jet_truthflav_save         = ROOT.vector("int")()
jet_DL1r_save              = ROOT.vector("float")()
classifier_save            = ROOT.vector("float")()
jet_pt_save                = ROOT.vector("float")()
jet_eta_save               = ROOT.vector("float")()
jet_phi_save               = ROOT.vector("float")()
jet_e_save                 = ROOT.vector("float")()

new_tree.Branch("topHadronOriginFlag",   topHOF_save)
new_tree.Branch("jet_isbtagged_DL1r_77", jet_isbtagged_DL1r_77_save)
new_tree.Branch("jet_truthflav",         jet_truthflav_save)
new_tree.Branch("jet_DL1r",              jet_DL1r_save)
new_tree.Branch(method_name,             classifier_save)
new_tree.Branch("jet_pt",                jet_pt_save)
new_tree.Branch("jet_eta",               jet_eta_save)
new_tree.Branch("jet_phi",               jet_phi_save)
new_tree.Branch("jet_e",                 jet_e_save)


# Loop over key of array - event numbers
for event_number in event_jet_jetinfo.keys():

  event_number = int(event_number)

  topHOF_save.clear()
  jet_isbtagged_DL1r_77_save.clear()
  jet_truthflav_save.clear()
  jet_DL1r_save.clear()
  classifier_save.clear()
  jet_pt_save.clear()
  jet_eta_save.clear()
  jet_phi_save.clear()
  jet_e_save.clear()

  # Loop over jets in the event - fill the "vector" values
  for jet_n in event_jet_jetinfo[event_number].keys():
    topHOF_save.push_back(int(event_jet_jetinfo[event_number][jet_n][0]))
    jet_isbtagged_DL1r_77_save.push_back(int(event_jet_jetinfo[event_number][jet_n][1]))
    jet_truthflav_save.push_back(int(event_jet_jetinfo[event_number][jet_n][2]))
    jet_DL1r_save.push_back(event_jet_jetinfo[event_number][jet_n][3])
    classifier_save.push_back(event_jet_jetinfo[event_number][jet_n][4])
    jet_pt_save.push_back(event_jet_jetinfo[event_number][jet_n][5])
    jet_eta_save.push_back(event_jet_jetinfo[event_number][jet_n][6])
    jet_phi_save.push_back(event_jet_jetinfo[event_number][jet_n][7])
    jet_e_save.push_back(event_jet_jetinfo[event_number][jet_n][8])


  # Fill the "single" values
  weight_save[0]    = event_lepinfo[event_number][0]
  met_save[0]       = event_lepinfo[event_number][1]
  el_pt_save[0]     = event_lepinfo[event_number][2]
  el_eta_save[0]    = event_lepinfo[event_number][3]
  el_phi_save[0]    = event_lepinfo[event_number][4]
  el_e_save[0]      = event_lepinfo[event_number][5]
  el_charge_save[0] = event_lepinfo[event_number][6]
  mu_pt_save[0]     = event_lepinfo[event_number][7]
  mu_eta_save[0]    = event_lepinfo[event_number][8]
  mu_phi_save[0]    = event_lepinfo[event_number][9]
  mu_e_save[0]      = event_lepinfo[event_number][10]
  mu_charge_save[0] = event_lepinfo[event_number][11]

  # Fill the new tree
  new_tree.Fill()

# Save the new file and close it
new_file.Write()
new_file.Close()
